pkg/controller/core.oam.dev/v1alpha2/application/assemble/assemble.go

```diff
@@ -0,0 +1,400 @@
+		oam.LabelAppComponent:         compName,
+		oam.LabelAppComponentRevision: compRevisionName,
+	}
+	// pass application's all labels to workload/trait
+	return util.MergeMapOverrideWithDst(Labels, am.appLabels)
+}
+
+// workload and trait both have these annotations
+func (am *AppManifests) setAnnotations(obj *unstructured.Unstructured) {
+	// pass application's all annotations
+	util.AddAnnotations(obj, am.appAnnotations)
+	// remove useless annotations for workload/trait
+	util.RemoveAnnotations(obj, []string{
+		oam.AnnotationAppRollout,
+		oam.AnnotationRollingComponent,
+		oam.AnnotationInplaceUpgrade,
+	})
+}
+
+func (am *AppManifests) setNamespace(obj *unstructured.Unstructured) {
+	// only set app's namespace when namespace is unspecified
+	// it's by design to set arbitrary namespace in render phase
+	if len(obj.GetNamespace()) == 0 {
+		obj.SetNamespace(am.appNamespace)
+	}
+}
+
+func (am *AppManifests) setOwnerReference(obj *unstructured.Unstructured) {
+	obj.SetOwnerReferences([]metav1.OwnerReference{*am.appOwnerRef})
+}
+
+func (am *AppManifests) assembleWorkload(comp *v1alpha2.Component, labels map[string]string) (*unstructured.Unstructured, error) {
+	compName := comp.Name
+	wl, err := util.RawExtension2Unstructured(&comp.Spec.Workload)
+	if err != nil {
+		return nil, errors.WithMessagef(err, "cannot convert raw workload in component %q", compName)
+	}
+	// use component name as workload name
+	// override the name set in render phase if exist
+	wl.SetName(compName)
+	am.setWorkloadLabels(wl, labels)
+	am.setAnnotations(wl)
+	am.setNamespace(wl)
+	am.setOwnerReference(wl)
+
+	workloadType := wl.GetLabels()[oam.WorkloadTypeLabel]
+	compDefinition := am.AppRevision.Spec.ComponentDefinitions[workloadType]
+	for _, wo := range am.WorkloadOptions {
+		if err := wo.ApplyToWorkload(wl, comp.DeepCopy(), compDefinition.DeepCopy()); err != nil {
+			klog.ErrorS(err, "Failed applying a workload option", "workload", klog.KObj(wl), "name", wl.GetName())
+			return nil, errors.Wrapf(err, "cannot apply workload option for component %q", compName)
+		}
+		klog.InfoS("Successfully apply a workload option", "workload", klog.KObj(wl), "name", wl.GetName())
+	}
+	klog.InfoS("Successfully assemble a workload", "workload", klog.KObj(wl), "APIVersion", wl.GetAPIVersion(), "Kind", wl.GetKind())
+	return wl, nil
+}
+
+func (am *AppManifests) setWorkloadLabels(wl *unstructured.Unstructured, commonLabels map[string]string) {
+	// add more workload-specific labels here
+	util.AddLabels(wl, map[string]string{oam.LabelOAMResourceType: oam.ResourceTypeWorkload})
+	util.AddLabels(wl, commonLabels)
+
+	/* NOTE a workload has these possible labels
+	   app.oam.dev/app-revision-hash: ce053923e2fb403f
+	   app.oam.dev/appRevision: myapp-v2
+	   app.oam.dev/component: mycomp
+	   app.oam.dev/name: myapp
+	   app.oam.dev/resourceType: WORKLOAD
+	   app.oam.dev/revision: mycomp-v2
+	   workload.oam.dev/type: kube-worker
+	*/
```

Comment:
let's make it as test case and remove it from here?

Reply:
sure, will do that. But I want to keep the labels key here to remind developers which labels are used because these labels come from different entities(app & xxDefiniton)

Reply:
We can keep it for now